File: src/mqttMidi/logDetail.ts
import { decodePitchBend, decodeSevenBit } from "./codec";
import { type ParsedMidiMessage, parseMidiMessage } from "./midiMessage";

function topicKind(topic: string): string | undefined {
  const parts = topic.split("/");
  const dirIndex = parts.findIndex((part) => part === "in" || part === "out");
  if (dirIndex < 0 || dirIndex + 1 >= parts.length) {
    return undefined;
  }
  return parts[dirIndex + 1];
}

/** Clock ticks flood the activity log; still bridge them, just don't display. */
export function shouldLogTraffic(topic: string): boolean {
  return topicKind(topic) !== "clock";
}

export function formatMidiDetail(message: ParsedMidiMessage): string {
  switch (message.kind) {
    case "noteon":
      return `noteon ch${message.channel} note${message.note} vel${message.velocity}`;
    case "noteoff":
      return `noteoff ch${message.channel} note${message.note} vel${message.velocity}`;
    case "cc":
      return `cc ch${message.channel} ctrl${message.controller} val${message.value}`;
    case "program":
      return `program ch${message.channel} prog${message.program}`;
    case "pitchbend":
      return `pitchbend ch${message.channel} val${message.value}`;
    case "sysex":
      return `sysex [${message.data.join(", ")}]`;
    case "system":
      return message.system;
  }
}

export function formatMqttDetail(topic: string, payload: Uint8Array | number[]): string {
  const bytes = payload instanceof Uint8Array ? payload : Uint8Array.from(payload);
  if (bytes.length === 0) {
    return topic;
  }

  const kind = topicKind(topic);
  try {
    switch (kind) {
      case "noteon":
      case "noteoff":
      case "cc":
      case "program":
        return `${topic} = ${decodeSevenBit(bytes)}`;
      case "pitchbend":
        return `${topic} = ${decodePitchBend(bytes)}`;
      case "sysex":
        return `${topic} = ${new TextDecoder().decode(bytes)}`;
      default:
        break;
    }
  } catch {
    // fall through to raw bytes
  }

  return `${topic} = [${Array.from(bytes).join(", ")}]`;
}

export function formatMidiToMqttDetail(
  midi: Uint8Array | number[],
  topic: string,
  payload: Uint8Array | number[],
): string {
  const parsed = parseMidiMessage(midi);
  const midiSide = parsed ? formatMidiDetail(parsed) : `[${Array.from(midi).join(", ")}]`;
  return `${midiSide} → ${formatMqttDetail(topic, payload)}`;
}

export function formatMqttToMidiDetail(
  topic: string,
  payload: Uint8Array | number[],
  midi: Uint8Array | number[],
): string {
  const parsed = parseMidiMessage(midi);
  const midiSide = parsed ? formatMidiDetail(parsed) : `[${Array.from(midi).join(", ")}]`;
  return `${formatMqttDetail(topic, payload)} → ${midiSide}`;
}

File: src-tauri/src/mqtt_midi/log_detail.rs
use super::codec::{decode_pitch_bend, decode_seven_bit};
use super::midi_message::{parse_midi_message, ParsedMidiMessage};

fn topic_kind(topic: &str) -> Option<&str> {
    let parts: Vec<&str> = topic.split('/').collect();
    let dir_index = parts.iter().position(|part| *part == "in" || *part == "out")?;
    parts.get(dir_index + 1).copied()
}

/// Clock ticks flood the activity log; still bridge them, just don't display.
pub fn should_log_traffic(topic: &str) -> bool {
    topic_kind(topic) != Some("clock")
}

pub fn format_midi_detail(message: &ParsedMidiMessage) -> String {
    match message {
        ParsedMidiMessage::NoteOn {
            channel,
            note,
            velocity,
        } => format!("noteon ch{channel} note{note} vel{velocity}"),
        ParsedMidiMessage::NoteOff {
            channel,
            note,
            velocity,
        } => format!("noteoff ch{channel} note{note} vel{velocity}"),
        ParsedMidiMessage::ControlChange {
            channel,
            controller,
            value,
        } => format!("cc ch{channel} ctrl{controller} val{value}"),
        ParsedMidiMessage::ProgramChange { channel, program } => {
            format!("program ch{channel} prog{program}")
        }
        ParsedMidiMessage::PitchBend { channel, value } => {
            format!("pitchbend ch{channel} val{value}")
        }
        ParsedMidiMessage::Sysex { data } => {
            let bytes = data
                .iter()
                .map(|byte| byte.to_string())
                .collect::<Vec<_>>()
                .join(", ");
            format!("sysex [{bytes}]")
        }
        ParsedMidiMessage::System(kind) => kind.as_str().to_string(),
    }
}

pub fn format_mqtt_detail(topic: &str, payload: &[u8]) -> String {
    if payload.is_empty() {
        return topic.to_string();
    }

    let formatted = match topic_kind(topic) {
        Some("noteon" | "noteoff" | "cc" | "program") => decode_seven_bit(payload)
            .ok()
            .map(|value| format!("{topic} = {value}")),
        Some("pitchbend") => decode_pitch_bend(payload)
            .ok()
            .map(|value| format!("{topic} = {value}")),
        Some("sysex") => String::from_utf8(payload.to_vec())
            .ok()
            .map(|json| format!("{topic} = {json}")),
        _ => None,
    };

    formatted.unwrap_or_else(|| {
        let bytes = payload
            .iter()
            .map(|byte| byte.to_string())
            .collect::<Vec<_>>()
            .join(", ");
        format!("{topic} = [{bytes}]")
    })
}

fn format_midi_bytes(midi: &[u8]) -> String {
    parse_midi_message(midi).map_or_else(
        || {
            let bytes = midi
                .iter()
                .map(|byte| byte.to_string())
                .collect::<Vec<_>>()
                .join(", ");
            format!("[{bytes}]")
        },
        |message| format_midi_detail(&message),
    )
}

pub fn format_midi_to_mqtt_detail(midi: &[u8], topic: &str, payload: &[u8]) -> String {
    format!(
        "{} → {}",
        format_midi_bytes(midi),
        format_mqtt_detail(topic, payload)
    )
}

pub fn format_mqtt_to_midi_detail(topic: &str, payload: &[u8], midi: &[u8]) -> String {
    format!(
        "{} → {}",
        format_mqtt_detail(topic, payload),
        format_midi_bytes(midi)
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn formats_midi_to_mqtt_as_midi_then_mqtt() {
        assert_eq!(
            format_midi_to_mqtt_detail(&[0x90, 60, 100], "remote/out/noteon/1/60", &[100]),
            "noteon ch1 note60 vel100 → remote/out/noteon/1/60 = 100"
        );
        assert_eq!(
            format_midi_to_mqtt_detail(&[0xb0, 7, 64], "remote/out/cc/1/7", &[64]),
            "cc ch1 ctrl7 val64 → remote/out/cc/1/7 = 64"
        );
    }

    #[test]
    fn formats_mqtt_to_midi_as_mqtt_then_midi() {
        assert_eq!(
            format_mqtt_to_midi_detail("remote/in/noteon/1/60", &[100], &[0x90, 60, 100]),
            "remote/in/noteon/1/60 = 100 → noteon ch1 note60 vel100"
        );
        assert_eq!(
            format_mqtt_to_midi_detail("remote/in/pitchbend/2", &[0, 64], &[0xe1, 0, 64]),
            "remote/in/pitchbend/2 = 8192 → pitchbend ch2 val8192"
        );
    }

    #[test]
    fn skips_clock_logging() {
        assert!(!should_log_traffic("remote/out/clock"));
        assert!(should_log_traffic("remote/out/start"));
        assert!(should_log_traffic("remote/out/noteon/1/60"));
    }

    #[test]
    fn formats_empty_mqtt_payload_without_equals() {
        assert_eq!(
            format_midi_to_mqtt_detail(&[0xfa], "remote/out/start", &[]),
            "start → remote/out/start"
        );
    }
}
